cdc/sink/buffer_sink.go

```diff
@@ -38,99 +41,126 @@ type bufferSink struct {
-	metricTotalRows := bufferSinkTotalRowsCountCounter.WithLabelValues(advertiseAddr, changefeedID)
+	state := runState{
+		metricFlushDuration:   flushRowChangedDuration.WithLabelValues(advertiseAddr, changefeedID, "Flush"),
+		metricEmitRowDuration: flushRowChangedDuration.WithLabelValues(advertiseAddr, changefeedID, "EmitRow"),
+		metricTotalRows:       bufferSinkTotalRowsCountCounter.WithLabelValues(advertiseAddr, changefeedID),
+	}
 	defer func() {
 		flushRowChangedDuration.DeleteLabelValues(advertiseAddr, changefeedID, "Flush")
 		flushRowChangedDuration.DeleteLabelValues(advertiseAddr, changefeedID, "EmitRow")
-		bufferChanSizeGauge.DeleteLabelValues(advertiseAddr, changefeedID)
 		bufferSinkTotalRowsCountCounter.DeleteLabelValues(advertiseAddr, changefeedID)
 	}()
+
 	for {
-		select {
-		case <-ctx.Done():
-			err := ctx.Err()
-			if err != nil && errors.Cause(err) != context.Canceled {
-				errCh <- err
-			}
+		keepRun, err := b.runOnce(ctx, &state)
+		if err != nil && errors.Cause(err) != context.Canceled {
+			errCh <- err
 			return
-		case drawback := <-b.drawbackChan:
-			b.bufferMu.Lock()
-			delete(b.buffer, drawback.tableID)
-			b.bufferMu.Unlock()
-			close(drawback.callback)
-		case flushEvent := <-b.flushTsChan:
-			b.bufferMu.Lock()
-			resolvedTs := flushEvent.resolvedTs
-			// find all rows before resolvedTs and emit to backend sink
-			for tableID, rows := range b.buffer {
-				i := sort.Search(len(rows), func(i int) bool {
-					return rows[i].CommitTs > resolvedTs
-				})
-				metricTotalRows.Add(float64(i))
-
-				start := time.Now()
-				err := b.Sink.EmitRowChangedEvents(ctx, rows[:i]...)
-				if err != nil {
-					b.bufferMu.Unlock()
-					if errors.Cause(err) != context.Canceled {
-						errCh <- err
-					}
-					return
-				}
-				dur := time.Since(start)
-				metricEmitRowDuration.Observe(dur.Seconds())
-
-				// put remaining rows back to buffer
-				// append to a new, fixed slice to avoid lazy GC
-				b.buffer[tableID] = append(make([]*model.RowChangedEvent, 0, len(rows[i:])), rows[i:]...)
+		}
+		if !keepRun {
+			return
+		}
+	}
+}
+
+func (b *bufferSink) runOnce(ctx context.Context, state *runState) (bool, error) {
+	batchSize, batch := 0, state.batch
+	push := func(event flushMsg) {
+		batch[batchSize] = event
+		batchSize++
+	}
+	select {
+	case <-ctx.Done():
+		return false, ctx.Err()
+	case drawback := <-b.drawbackChan:
+		b.bufferMu.Lock()
+		delete(b.buffer, drawback.tableID)
+		b.bufferMu.Unlock()
+		close(drawback.callback)
+	case event := <-b.flushTsChan:
+		push(event)
+	RecvBatch:
+		for batchSize < maxFlushBatchSize {
+			select {
+			case event := <-b.flushTsChan:
+				push(event)
+			default:
+				break RecvBatch
 			}
+		}
+	}
+
+	b.bufferMu.Lock()
+	startEmit := time.Now()
+	// find all rows before resolvedTs and emit to backend sink
+	for i := 0; i < batchSize; i++ {
+		tableID, resolvedTs := batch[i].tableID, batch[i].resolvedTs
```

Comment:
if there are many flush events in `batch` for one table, does it help to improve the performance to find the max resovedTs per table first,  do not need to search and flush multi times?